source/Timer.cpp

```diff
@@ -0,0 +1,322 @@
+	result.proximityCenter = proximityCenter;
+	result.proximityCenters = proximityCenters;
+	result.closeTo = closeTo;
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	result.action = action.Instantiate(subs, origin, jumps, payload);
+	result.resetAction = resetAction.Instantiate(subs, origin, jumps, payload);
+
+	if(rand >= 1)
+		result.timeToWait = base + Random::Int(rand);
+	else
+		result.timeToWait = base;
+
+	return result;
+}
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
+
+
+
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		timeToWait = base + Random::Int(rand);
+		if(repeatReset || !resetFired)
+		{
+			resetAction.Do(player, ui, mission);
+			resetFired = true;
+		}
+		isActive = false;
+	}
+}
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui)
+{
+	if(isComplete)
+		return;
+	if(!player.Flagship())
+		return;
+	if((system && player.Flagship()->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(player.Flagship()->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM, player, ui);
+		return;
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!player.Flagship()->IsThrusting() && !player.Flagship()->IsSteering()
+						&& !player.Flagship()->IsReversing());
+		for(const Hardpoint &weapon : player.Flagship()->Weapons())
+			shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui);
+			return;
+		}
+	}
+	if(requireUncloaked)
+	{
+		double cloak = player.Flagship()->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui);
+			return;
+		}
+	}
+	if(proximity > 0.)
+	{
+		bool inProximity = false;
+		if(proximityCenter || !proximityCenters.IsEmpty())
+		{
+			for(const StellarObject &proximityObject : system->Objects())
+				if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+					proximityCenters.Matches(proximityObject.GetPlanet())))
+				{
+					double dist = player.Flagship()->Position().Distance(proximityObject.Position());
+					if((closeTo && dist <= proximity) || (!closeTo && dist >= proximity))
+					{
+						inProximity = true;
+						break;
+					}
+				}
+		}
+		if(!inProximity)
+		{
+			ResetOn(Timer::ResetCondition::LEAVE_ZONE, player, ui);
+			return;
+		}
+	}
+	isActive = true;
+	++timeElapsed;
+	if(timeElapsed >= timeToWait)
```

Comment:
You could just increment in the comparison. The prefixed ++ increments and then compares (whereas if it were `var++ >= value` it would compare and then increment).
```suggestion
	if(++timeElapsed >= timeToWait)
```